server/src/state/file.rs

```diff
@@ -0,0 +1,193 @@
+    }
+}
+
+#[async_trait]
+impl State for FileState {
+    async fn init(&self) -> Result<Vec<StateEntry>, IggyError> {
+        if !Path::new(&self.path).exists() {
+            info!("State file does not exist, creating a new one");
+            self.persister.overwrite(&self.path, &[]).await?;
+        }
+
+        let entries = self.load_entries().await?;
+        let entries_count = entries.len() as u64;
+        self.entries_count.store(entries_count, Ordering::SeqCst);
+        if entries_count == 0 {
+            self.current_index.store(0, Ordering::SeqCst);
+        } else {
+            let last_index = entries[entries_count as usize - 1].index;
+            self.current_index.store(last_index, Ordering::SeqCst);
+        }
+
+        return Ok(entries);
+    }
+
+    async fn load_entries(&self) -> Result<Vec<StateEntry>, IggyError> {
+        if !Path::new(&self.path).exists() {
+            return Err(IggyError::StateFileNotFound);
+        }
+
+        let file = file::open(&self.path).await?;
+        let file_size = file.metadata().await?.len();
+        if file_size == 0 {
+            info!("State file is empty");
+            return Ok(Vec::new());
+        }
+
+        info!(
+            "Loading state, file size: {}",
+            IggyByteSize::from(file_size).as_human_string()
+        );
+        let mut entries = Vec::new();
+        let mut total_size: u64 = 0;
+        let mut reader = BufReader::with_capacity(BUF_READER_CAPACITY_BYTES, file);
+        let mut current_index = 0;
+        let mut entries_count = 0;
+        loop {
+            let index = reader.read_u64_le().await?;
+            if entries_count > 0 && index != current_index + 1 {
+                error!(
+                    "State file is corrupted, expected index: {}, got: {}",
+                    current_index + 1,
+                    index
+                );
+                return Err(IggyError::StateFileCorrupted);
+            }
+
+            current_index = index;
+            entries_count += 1;
+            let term = reader.read_u64_le().await?;
+            let leader_id = reader.read_u32_le().await?;
+            let version = reader.read_u32_le().await?;
+            let flags = reader.read_u64_le().await?;
+            let timestamp = IggyTimestamp::from(reader.read_u64_le().await?);
+            let user_id = reader.read_u32_le().await?;
+            let code = reader.read_u32_le().await?;
+            let payload_length = reader.read_u32_le().await? as usize;
+            let mut payload = BytesMut::with_capacity(payload_length);
+            payload.put_bytes(0, payload_length);
+            reader.read_exact(&mut payload).await?;
+            let context_length = reader.read_u32_le().await? as usize;
+            let mut context = BytesMut::with_capacity(context_length);
+            context.put_bytes(0, context_length);
+            reader.read_exact(&mut context).await?;
+            let entry = StateEntry {
```

Comment:
Maybe create a `new` method for this instead of using constructor

Reply:
Ok